engine/src/main/java/org/terasology/rendering/dag/RenderGraph.java

```diff
@@ -15,44 +15,68 @@
  */
 package org.terasology.rendering.dag;
 
+import com.google.common.collect.HashMultimap;
 import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Multimap;
 import org.terasology.engine.SimpleUri;
 
+import java.util.ArrayList;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
 
 /**
  * TODO: Add javadocs
  */
 public class RenderGraph { // TODO: add extends DirectedAcyclicGraph<Node>
-    private List<Node> nodes;
-
-    public RenderGraph() {
-        nodes = Lists.newArrayList();
-    }
+    private Map<SimpleUri, Node> nodes = Maps.newHashMap();
+    private Multimap<Node, Node> nodeEdges = HashMultimap.create();
 
     public SimpleUri addNode(Node node, String suggestedUri) {
-        nodes.add(node);
-
         // TODO: make sure URIs are actually unique: if "myModule:blur" is present the node gets the uri "myModule:blur2" instead.
         // TODO: make sure the namespace in the uri is engine-assigned, so that only engine nodes can have the "engine:" namespace - everything else gets the namespace of the module.
         SimpleUri nodeUri = new SimpleUri("engine:" + suggestedUri);
-        node.setUri(nodeUri);
+
+        nodes.put(nodeUri, node);
+
         return nodeUri;
     }
 
-    public Node findNode(SimpleUri nodeUri) {
-        for (Node node: nodes) {
-            if (node.getUri().equals(nodeUri)) {
-                return node;
-            }
+    public boolean removeNode(SimpleUri nodeUri) {
+        if (nodeEdges.containsKey(nodeUri)) {
+            return false;
         }
 
-        return null;
+        nodes.remove(nodeUri);
+
+        return true; // TODO: What if nodeUri does not refer to a Node?
+    }
+
+    public Node findNode(SimpleUri nodeUri) {
+        return nodes.get(nodeUri);
+    }
+
+    public void addEdge(Node node1, Node node2) {
```

Comment:
The disparity in name (add/remove edge vs connect/disconnect) was an oversight on my part, and is fixed now.

Reply:
> Again we could return a boolean as a "success status" (and log a warning)
Why exactly would we want to log a warning?

Reply:
Correct: by returning false the caller of the method is informed of the failure and has the chance to react to it. But that doesn't say anything about what the failure was and a log warning would disclose that (i.e., the 2nd argument was null). 

That been said, as I think about it, I wonder if we should generate exceptions instead, so that the code can react to each individual problem that might arise. In this context we wouldn't return anything: if no exception is thrown the call was successful. 

What do you think?

Reply:
I'm of the opinion that exceptions are the way to go.

That being said, I still want to return a boolean, to indicate the success. What if the arguments were not null, but they were already connected? This call would fail in that case. And since I really don't think module developers are likely to check the return value of `connect()` , I also added a warning (similar to how addDesiredStateChange logs a warning when we try to add default SCs).

Check out the last commit.